File: computing/interview/design-patterns/creational/abstract-factory/cpp/abstract_factory.cpp
#include <iostream>
#include <memory>

struct Button {
    virtual ~Button() = default;
    virtual void render() const = 0;
};

struct Checkbox {
    virtual ~Checkbox() = default;
    virtual void toggle() const = 0;
};

struct GuiFactory {
    virtual ~GuiFactory() = default;
    virtual std::unique_ptr<Button> create_button() const = 0;
    virtual std::unique_ptr<Checkbox> create_checkbox() const = 0;
};

class WindowsButton : public Button {
public:
    void render() const override { std::cout << "Rendering Windows button\n"; }
};

class WindowsCheckbox : public Checkbox {
public:
    void toggle() const override { std::cout << "Toggling Windows checkbox\n"; }
};

class WindowsFactory : public GuiFactory {
public:
    std::unique_ptr<Button> create_button() const override { return std::make_unique<WindowsButton>(); }
    std::unique_ptr<Checkbox> create_checkbox() const override { return std::make_unique<WindowsCheckbox>(); }
};

class MacButton : public Button {
public:
    void render() const override { std::cout << "Rendering macOS button\n"; }
};

class MacCheckbox : public Checkbox {
public:
    void toggle() const override { std::cout << "Toggling macOS checkbox\n"; }
};

class MacFactory : public GuiFactory {
public:
    std::unique_ptr<Button> create_button() const override { return std::make_unique<MacButton>(); }
    std::unique_ptr<Checkbox> create_checkbox() const override { return std::make_unique<MacCheckbox>(); }
};

class Application {
public:
    explicit Application(std::unique_ptr<GuiFactory> factory) : factory_(std::move(factory)) {}

    void paint_ui() const {
        auto button = factory_->create_button();
        auto checkbox = factory_->create_checkbox();
        button->render();
        checkbox->toggle();
    }

private:
    std::unique_ptr<GuiFactory> factory_;
};

int main() {
    Application windows_app(std::make_unique<WindowsFactory>());
    windows_app.paint_ui();

    Application mac_app(std::make_unique<MacFactory>());
    mac_app.paint_ui();
}

File: computing/interview/design-patterns/creational/abstract-factory/python/abstract_factory.py
from __future__ import annotations
from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def render(self) -> None:
        ...


class Checkbox(ABC):
    @abstractmethod
    def toggle(self) -> None:
        ...


class GuiFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        ...

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        ...


class WindowsButton(Button):
    def render(self) -> None:
        print("Rendering Windows button")


class WindowsCheckbox(Checkbox):
    def toggle(self) -> None:
        print("Toggling Windows checkbox")


class WindowsFactory(GuiFactory):
    def create_button(self) -> Button:
        return WindowsButton()

    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()


class MacButton(Button):
    def render(self) -> None:
        print("Rendering macOS button")


class MacCheckbox(Checkbox):
    def toggle(self) -> None:
        print("Toggling macOS checkbox")


class MacFactory(GuiFactory):
    def create_button(self) -> Button:
        return MacButton()

    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()


class Application:
    def __init__(self, factory: GuiFactory) -> None:
        self._factory = factory

    def paint_ui(self) -> None:
        button = self._factory.create_button()
        checkbox = self._factory.create_checkbox()
        button.render()
        checkbox.toggle()


def main() -> None:
    Application(WindowsFactory()).paint_ui()
    Application(MacFactory()).paint_ui()


if __name__ == "__main__":
    main()
